File: 프로그래머스/lv2/81302. 거리두기 확인하기/거리두기 확인하기.js
function BFS(graph2,a,b){
    let graph3 = []
    for(let x of graph2){
        graph3.push([...x])
    }
    let dx = [-1,1,0,0]
    let dy = [0,0,-1,1]
    graph3[a][b] = 0
    let queue = [[a,b]]
    let count = 0
    while(queue.length!==0){
        let [x,y] = queue.shift()
        for(let i=0;i<4;i++){
            let nx = x+dx[i]
            let ny = y+dy[i]
            if(nx<0 || ny<0 || nx>=5 || ny>=5 || graph3[nx][ny] === 'X' || !isNaN(graph3[nx][ny])){
                continue
            }
            if(graph3[nx][ny] === 'P'){
                if(Math.abs(nx-a)+Math.abs(ny-b) <=2){
                    if(graph3[x][y]<=1) {
                        return false
                    }
                    
                }
                else{
                    queue.push([nx,ny])
                    graph3[nx][ny] = graph3[x][y]+1;
                    continue
                }
            }
            graph3[nx][ny] = graph3[x][y]+1;
            queue.push([nx,ny])
        }
    }
    return true
}

function solution(places) {
    var answer = [];
//     갈수있는지없는지 보면됨
//     맨해튼거리계산은 좌표랑 절댓값만 쓰면됨
    for(let x of places){
        let result = 1
        // console.log(x)
//     모든점 돌면서 플레이어가있는 곳에서 출발.
        let graph = []
        for(let q of x){
            graph.push(q.split(""))
        }
        (()=>{
        for(let x=0;x<5;x++){
            for(let y=0;y<5;y++){
                if(graph[x][y] === "P"){
//                     탐색후, 결과에따라 result 0으로 변경
                    if(!BFS(graph,x,y)){
                        result = 0
                        return;
                    }
                }
            }
        }
        })()
        answer.push(result)
    }
    return answer;
}

File: 프로그래머스/lv2/81302. 거리두기 확인하기/거리두기 확인하기.py
def solution(places):
    answer = []
    #거리2까지만 보자
    #후보군은 12개, 그중에서 맨해튼거리가 1인애들은 어쩔수없음. 안되는거임
    #맨해튼거리가 2인애들은 경우에따라 막혀있을수도있음
    #상하로 거리2
    dx1=[-2,2,0,0]
    dy1=[0,0,-2,2]
    #대각으로 거리2
    dx2=[-1,-1,1,1]
    dy2=[1,-1,1,-1]
    #거리1
    dx3=[-1,1,0,0]
    dy3=[0,0,-1,1]
    
    for i in places:
        graph = i
        coordinates = []
        flag = True
        for k in range(5):
            for u in range(5):
                if graph[k][u] == 'P':
                    coordinates.append((k,u))
        for k in coordinates:
            x,y = k
            for u in range(4):
                nx = x+dx1[u]
                ny = y+dy1[u]
                if nx<0 or ny<0 or nx>=5 or ny>=5:
                    continue
                if graph[nx][ny] == 'P':
                    if graph[int((x+nx)/2)][int((y+ny)/2)] == 'X':
                        continue
                    else:
                        flag = False
                        break
            if not flag:
                break
                        
            for u in range(4):
                nx = x+dx2[u]
                ny = y+dy2[u]
                if nx<0 or ny<0 or nx>=5 or ny>=5:
                    continue
                if graph[nx][ny] == 'P':
                    if graph[nx][y] and graph[x][ny] == 'X':
                        continue
                    else:
                        flag = False
                        break
                        
            if not flag:
                break
            
            for u in range(4):
                nx = x+dx3[u]
                ny = y+dy3[u]
                if nx<0 or ny<0 or nx>=5 or ny>=5:
                    continue
                if graph[nx][ny] == 'P':
                    flag = False
                    break
                        
            if not flag:
                break
                
        if flag:
            answer.append(1)
        else:
            answer.append(0)
            
    return answer